torch/distributed/nn/functional.py

```diff
@@ -365,21 +365,16 @@ def forward(ctx, output_tensor, input_tensor, group):
     @staticmethod
     # pyrefly: ignore [bad-override]
     def backward(ctx, grad_output):
-        if dist.get_backend(group=ctx.group) is dist.Backend.NCCL:
-            world_size = dist.get_world_size(group=ctx.group)
-            out_size = list(grad_output.size())
-            if out_size[0] % world_size != 0:
-                raise RuntimeError(
-                    f"Tensor with dimensions: {out_size} does "
-                    f"not have first dimension divisible by world_size: {world_size}"
-                )
-            out_size[0] = out_size[0] // dist.get_world_size(group=ctx.group)
-            gx = torch.empty(
-                out_size, device=grad_output.device, dtype=grad_output.dtype
+        world_size = dist.get_world_size(group=ctx.group)
```

Comment:
This change enables all distributed backends, including NCCL, XCCL, and RCCL. So we need to verify Rocm as well. By the way, pls. add a regression test to validate this PR.